### ISSUE TITLE
[BUG] Grad norm value differs significantly across DeepSpeed version

### ISSUE THREAD

**Issue by npurson** [deepspeedai/DeepSpeed#7749]
**Describe the bug**

I observed a significant discrepancy in `grad_norm` values when upgrading from DeepSpeed 0.16.4 to 0.18.3. Under **identical training configurations** (same model, same hyperparameters, same script), the reported `grad_norm` drops by orders of magnitude (from ~11-15 to ~0.2).

I am finetuning [Qwen2.5_VL](https://github.com/QwenLM/Qwen3-VL/blob/main/qwen-vl-finetune/scripts/sft.sh) with DeepSpeed ZeRO-2. When I am using DeepSpeed 0.16.4, the training log is:

```
{'loss': 1.7677, 'grad_norm': 11.14142894744873, 'learning_rate': 4.025044722719142e-07, 'epoch': 0.0}                                                                                                                                
{'loss': 2.2447, 'grad_norm': 11.526385307312012, 'learning_rate': 4.1144901610017893e-07, 'epoch': 0.0}                                                                                                                              
{'loss': 1.774, 'grad_norm': 12.301006317138672, 'learning_rate': 4.203935599284437e-07, 'epoch': 0.0}                                                                                                                                
{'loss': 2.0844, 'grad_norm': 15.470754623413086, 'learning_rate': 4.2933810375670843e-07, 'epoch': 0.0}                                                                                                                              
{'loss': 2.0122, 'grad_norm': 14.170842170715332, 'learning_rate': 4.3828264758497323e-07, 'epoch': 0.0}                                                                                                                              
{'loss': 2.106, 'grad_norm': 12.281550407409668, 'learning_rate': 4.47227191413238e-07, 'epoch': 0.0}                                                                                                                                                                                                                                                           
  0%|▎                                                                                                                                                                                          | 59/37254 [06:18<66:08:28,  6.40s/it]
```

After upgrading to DeepSpeed 0.18.3, the training log is:

```
{'loss': 1.958, 'grad_norm': 0.21968881785869598, 'learning_rate': 0.0, 'epoch': 0.0}                                                                                                                                                 
{'loss': 1.7761, 'grad_norm': 0.1559031903743744, 'learning_rate': 8.94454382826476e-09, 'epoch': 0.0}                                                                                                                                
  0%|                                                                                                                                                                                            | 2/37254 [00:14<68:36:25,  6.63s/it]
```

Is this the expected behavior or a bug?

**ds_report output**

```
DeepSpeed C++/CUDA extension op report
--------------------------------------------------
NOTE: Ops not installed will be just-in-time (JIT) compiled at
      runtime if needed. Op compatibility means that your system
      meet the required dependencies to JIT install the op.
--------------------------------------------------
JIT compiled ops requires ninja
ninja .................. [OKAY]
--------------------------------------------------
op name ................ installed .. compatible
--------------------------------------------------
 [WARNING]  async_io requires the dev libaio .so object and headers but these were not found.
 [WARNING]  async_io: please install the libaio-dev package with apt
 [WARNING]  If libaio is already installed (perhaps from source), try setting the CFLAGS and LDFLAGS environment variables to where it can be found.
async_io ............... [NO] ....... [NO]
fused_adam ............. [NO] ....... [OKAY]
cpu_adam ............... [NO] ....... [OKAY]
cpu_adagrad ............ [NO] ....... [OKAY]
cpu_lion ............... [NO] ....... [OKAY]
dc ..................... [NO] ....... [OKAY]
 [WARNING]  Please specify the CUTLASS repo directory as environment variable $CUTLASS_PATH
evoformer_attn ......... [NO] ....... [NO]
 [WARNING]  FP Quantizer is using an untested triton version (3.4.0), only 2.3.(0, 1) and 3.0.0 are known to be compatible with these kernels
fp_quantizer ........... [NO] ....... [NO]
fused_lamb ............. [NO] ....... [OKAY]
fused_lion ............. [NO] ....... [OKAY]
 [WARNING]  gds requires the dev libaio .so object and headers but these were not found.
 [WARNING]  gds: please install the libaio-dev package with apt
 [WARNING]  If libaio is already installed (perhaps from source), try setting the CFLAGS and LDFLAGS environment variables to where it can be found.
gds .................... [NO] ....... [NO]
transformer_inference .. [NO] ....... [OKAY]
inference_core_ops ..... [NO] ....... [OKAY]
cutlass_ops ............ [NO] ....... [OKAY]
quantizer .............. [NO] ....... [OKAY]
ragged_device_ops ...... [NO] ....... [OKAY]
ragged_ops ............. [NO] ....... [OKAY]
random_ltd ............. [NO] ....... [OKAY]
 [WARNING]  sparse_attn requires a torch version >= 1.5 and < 2.0 but detected 2.8
 [WARNING]  using untested triton version (3.4.0), only 1.0.0 is known to be compatible
sparse_attn ............ [NO] ....... [NO]
spatial_inference ...... [NO] ....... [OKAY]
transformer ............ [NO] ....... [OKAY]
stochastic_transformer . [NO] ....... [OKAY]
utils .................. [NO] ....... [OKAY]
--------------------------------------------------
DeepSpeed general environment info:
torch install path ............... ['/home/users/xxx/envs/miniconda3/envs/xxx/lib/python3.12/site-packages/torch']
torch version .................... 2.8.0+cu128
deepspeed install path ........... ['/home/users/xxx/envs/miniconda3/envs/xxx/lib/python3.12/site-packages/deepspeed']
deepspeed info ................... 0.18.3, unknown, unknown
torch cuda version ............... 12.8
torch hip version ................ None
nvcc version ..................... 11.8
deepspeed wheel compiled w. ...... torch 0.0, cuda 0.0
shared memory (/dev/shm) size .... 503.76 GB
```

**System info:**

 - OS: Ubuntu 22.04
 - GPU count and types 1x 4090 for debug and 8x 5090 for training.
 - Python version: both 3.10 & 3.12
 - Any other relevant info about your setup

**Launcher context**

torchrun

**Comment by kimmeoungjun** [comment]
In your logs, the learning_rate values differ between versions.
Are these logs captured at the same optimizer step, or could the logging point have changed?

**Comment by npurson** [comment]
> In your logs, the learning_rate values differ between versions.
> Are these logs captured at the same optimizer step, or could the logging point have changed?

While the log snippets I posted above are not from the exact same step, I have confirmed that this significant discrepancy in grad_norm persists consistently throughout the entire training process, and the order of grad_norm magnitude difference remains the same at identical optimizer steps.

BTW, I am using the codebase of https://github.com/LaVi-Lab/VG-LLM

**Comment by tdrussell** [comment]
I raise issue #7773 earlier today, and then accidentally stumble upon your post. Maybe your problem is related to the same issue I see (which is fixed starting in 0.18.3). You say you are using Zero-2, which I thought wasn't compatible with pipeline parallelism, so it might not be exactly the same issue. But it could be related. If you are using gradient accumulation steps > 1, you might check if that is affecting the grad norm value differently for Deepspeed 0.18.3 vs older versions.

**Comment by alanhuangyoo** [comment]
I measured what current master actually reports, so at least the space of explanations narrows. Short version: on master the reported `grad_norm` does not depend on gradient accumulation, on the ZeRO stage, or on bf16 vs fp32. So whatever changed between 0.16.4 and 0.18.3 for you, it is not one of those three on its own.

## Setup

A 2-layer MLP, 2 GPUs, `lr=0` and `gradient_clipping=1e9` so the weights never move and every configuration sees the identical gradient. The same 16 samples per rank per optimizer step in every row — one micro-batch of 16 with `gas=1`, four of 4 with `gas=4` — so the mean gradient is the same by construction and the reported norm has to be too.

## fp32

```
stage=1 gas=1  grad_norm=0.07462712      stage=2 gas=1  0.07462712      stage=3 gas=1  0.07462712
stage=1 gas=2  grad_norm=0.07462712      stage=2 gas=2  0.07462712      stage=3 gas=2  0.07462712
stage=1 gas=4  grad_norm=0.07462712      stage=2 gas=4  0.07462712      stage=3 gas=4  0.07462712
```

Identical to the last digit across all nine.

## bf16

```
stage=0  grad_norm=0.07461283      stage=3  grad_norm=0.07461282
```

Same value to ~2e-5 relative, which is bf16 rounding on the gradient itself.

## What that leaves

The ~50x drop you see is not the accumulation accounting and not the stage. The remaining candidates, in the order I would check them:

1. **What the number is a norm *of*.** `_global_grad_norm` is the norm of the reduced, accumulated gradient before the optimizer applies weight decay or an update. If 0.16.4 was reporting something else — a pre-reduction norm, or a norm still carrying a scale — then 0.2 is the correct value and the old one was inflated. @tdrussell's #7773 is exactly that shape of bug, and it landed in 0.18.3.
2. **Rank counting.** A norm summed over ranks that should have been summed over partitions gives `sqrt(world_size)`, which is 2-4x, not 50x — so this alone does not explain it either.
3. **Frozen parameters.** Qwen2.5-VL finetuning usually freezes the vision tower. If the two versions disagree about whether frozen or `None`-gradient parameters enter the norm, the magnitudes move.

To narrow it further I would need three numbers from your run that are not in the thread yet: `world_size`, `gradient_accumulation_steps`, and whether you set `gradient_clipping`. If you can also print `sum(p.grad.float().norm()**2 for p in model.parameters() if p.grad is not None) ** 0.5` right before `engine.step()` on both versions, that separates "the gradient changed" from "the reporting changed", which is the fork everything else hangs on.

The scripts above are ~40 lines; happy to post them if you want to run the same comparison on 0.16.4.
